### PR TITLE
Extended Data Model

### PR DESCRIPTION
**Pull request by avihut**
Most of the work here went towards making the data model better. The data always flows inwards from Facebook. All of the data is saved locally, and only the most basic information that is required for payments is saved into Parse.

Also added RSVP information that will be used for the detail display of the event information. Currently it's not even fetched from Facebook yet.

Also added a display for empty state and pull to refresh.
### Refresh Animation Now Displayed when the Application Opens

refresh on open
### Pull to Refresh Will Change to Empty State of All Events were Deleted

refresh to empty state
### Clicking the Big Refresh Button in the Empty State will Reveal the List of Events

refresh from empty to single
### Pulling the List to Refresh will Load More Information if Available

refresh for more

### CODE REVIEW

File: Event.swift [hunk 1]
```diff
@@ -11,15 +11,21 @@ import Parse
 
 class Event: PFObject, PFSubclassing, Comparable {
     
-    convenience init(fbId: String, ownerFbId: String, name: String) {
+    static func findOrCreateBlocking(fbId: String) -> Event {
+        let findEventQuery = PFQuery(className: parseClassName()).whereKey("fbId", equalTo: fbId)
+        if let events = findEventQuery.findObjects() as? Events, event = events.first {
```
Comment (yonbergman) [inline Event.swift]: i'm pretty sure PFQuery has a method that returns just the first result, no?

File: PSix/Array.swift [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+//
+//  Created by Turzion, Avihu on 4/29/15.
+//  Copyright (c) 2015 PayPal. All rights reserved.
+//
+
+import Foundation
+
+extension Array {
+    
+    static func filterNils(array: [T?]) -> [T] {
```
Comment (yonbergman) [inline PSix/Array.swift:13]: maybe bring in dollar? the swift library?

File: PSix/EventCell.swift [hunk 3]
```diff
@@ -32,7 +32,12 @@ class EventCell: UITableViewCell {
                 eventDateMonthLabel.text = event.startTime?.monthShortName.uppercaseString
                 eventDateDayLabel.text = event.startTime?.dayDoubleDigit
                 eventDateDayAndTime.text = EventPresenter.getDayHourOfStartConsideringWidth(event, boxWidth: eventDateDayAndTime.bounds.width, font: eventDateDayAndTime.font)
-                paymentCollectionStatus.text = EventPresenter.getPaymentStatus(event)
+                let paymentCollectionStatus = EventPresenter.getPaymentStatus(event)
```
Comment (yonbergman) [inline PSix/EventCell.swift]: I don't like the fact that you have a presenter which isn't really acting like a presenter, more like a set of utility methods.

File: PSix/EventsListViewController.swift [hunk 4]
```diff
@@ -47,8 +50,23 @@ class EventsListViewController: UIViewController, UITableViewDataSource, UITable
         eventsListTable.delegate = self
     }
     
+    func refreshData() {
+        updateUserEvents()
+    }
+    
+    private func getUserOnboard() {
```
Comment (yonbergman) [inline PSix/EventsListViewController.swift:57]: presentUserOnboarding

File: PSix/EventsListViewController.swift [hunk 5]
```diff
@@ -63,4 +81,11 @@ class EventsListViewController: UIViewController, UITableViewDataSource, UITable
         return cell
     }
     
+    @IBAction func refreshForNewEvents() {
+        noEventsRefreshSpinner.startAnimating()
+        updateUserEvents() { [unowned self] () -> Void in
```
Comment (yonbergman) [inline PSix/EventsListViewController.swift]: no need for -> void

File: Event.swift (hunk 1 above)
Comment (avihut) [inline Event.swift]: Fixed

File: Event.swift [hunk 6]
```diff
@@ -31,19 +37,49 @@ class Event: PFObject, PFSubclassing, Comparable {
...
-    @NSManaged var endTime: NSDate?
+    
+    var name: String?
+    var eventDescription: String?
+    var location: String?
+    var startTime: NSDate?
+    var endTime: NSDate?
     var coverImageUrl: NSURL?
     
+    var _paymentsRelation: PFRelation?
```
Comment (assafgelber) [inline Event.swift]: If I understand what you're doing here, I think you could re-write as:

``` swift
lazy var paymentsRelation = {
  self.relationForKey("payments")
}
```

File: RSVP.swift [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    
+    static func filterAttending(rsvps: RSVPs) -> RSVPs {
+        return filterRsvpsByStatus(rsvps, status: .Attending)
+    }
+    
+    static func filterMaybe(rsvps: RSVPs) -> RSVPs {
+        return filterRsvpsByStatus(rsvps, status: .Maybe)
+    }
+    
+    static func filterDeclined(rsvps: RSVPs) -> RSVPs {
```
Comment (assafgelber) [inline RSVP.swift]: I think having 3 different methods for this is too much. How about:

``` swift
static func filter(rsvps: RSVPs, byStatus status: RsvpStatus) -> RSVPs
```

File: PSix/Array.swift (hunk 2 above)
Comment (avihut) [inline PSix/Array.swift:13]: After wasting an hour debugging cocoa pods for this, I am opting out. I definitely see the advantage of Dollar, but I have no idea how to integrate it.

Comment (assafgelber) [inline PSix/Array.swift:13]: Did you:
- [x] Add `pod 'Dollar'` to your podfile
- [x] Add `use_frameworks!` to your podfile
- [x] Use cocoapods `0.36.1` (Yon told us all to use `0.36.0`, so you may have to use rvm)

?

File: PSix/EventsListViewController.swift (hunk 4 above)
Comment (avihut) [inline PSix/EventsListViewController.swift:57]: Fixed

File: PSix/EventsListViewController.swift (hunk 5 above)
Comment (avihut) [inline PSix/EventsListViewController.swift]: I do, because I declare `[unowned self]`

Comment (avihut) [inline PSix/EventsListViewController.swift]: I stand corrected

File: PSix/Array.swift (hunk 2 above)
Comment (avihut) [inline PSix/Array.swift:13]: I give up....
giphy

File: PSix/EventCell.swift (hunk 3 above)
Comment (avihut) [inline PSix/EventCell.swift]: Fixed. Refactored it to look like it's used in Mokojin.

File: Event.swift (hunk 6 above)
Comment (avihut) [inline Event.swift]: Indeed! Thanks :)

File: RSVP.swift (hunk 7 above)
Comment (avihut) [inline RSVP.swift]: Fixed

**Pull request closed without merging**